### DISCUSSION TITLE
Finit State-Machine with Berry

### DISCUSSION THREAD

**Discussion by wjohn007** [arendst/Tasmota#22705] [Show and tell]
In one of my projects the processes became very complex, so I decided to use a finit state machine.

The following example application shows the states and transitions.

![fsm](https://github.com/user-attachments/assets/b24359ce-d7e3-4aea-8dff-6521d3594f4d)

This graphic was also created by Berry. It shows the current state of the state machine. (Mermaid renders the graphic)

The source code and the example application can be found in the [Git repo](https://github.com/wjohn007/Berry-Finit-State-Machine)

**Comment by GEAG** [comment]
This destructor would probably never be called by the Berry GC, unless the driver has previously already been removed, because the tasmota object holds a reference on the object until it is removed. 
  
```
  # destructor
    def deinit()
        var cproc="deinit"
        tasmota.remove_driver(self)
        self.info(cproc,"done")
    end 
```
    
A "week reference" would make sense here. Does anybody know how week references could can be realized with Berry ?

**Comment by s-hadinger** [comment]
There is no weak reference in Berry.

Indeed the destructor above will never be called if it's still registered as a driver in Tasmota. The instance needs to call `tasmota.remove_driver(self)` when it has finished its work. Only later will `deinit` be called.

Be very careful, any exception or anything going wrong in `deinit` will most certainly cause a crash. You are not allowed to allocated any new memory in `deinit`. Its only purpose is to free resources outside of Berry (like files open or C resources).
